sql: fix integer overflow in built-in functions

ImeevMA opened the pull request:
This patch replaces the type for some int and uint32_t values with size_t to avoid problems with integer overflow.

Closes tarantool/security#119

NO_DOC=bugfix

Gerold103 commented on src/box/sql/func.c:
```diff
@@ -671,6 +689,7 @@ func_substr_octets(struct sql_context *ctx, int argc, const struct Mem *argv)
 	}
 	char *str = &argv[0].z[start];
 	uint64_t len = MIN(size - start, length);
+	assert(len <= SQL_MAX_LENGTH);
 	ctx->is_aborted = is_str ? mem_copy_str(ctx->pOut, str, len) != 0 :
```
Maybe just add this assert to `mem_copy_str()`, `mem_copy_bin()`, and whatever other mem methods require it. It doesn't look right to copy-paste this assert before each usage of mem copy methods.

ImeevMA replied:
Hi! Thank you for the review! I have added such asserts to all of these functions. Also, I've added an `if` where it's possible for the length to be greater than SQL_MAX_LENGTH, and an assert where it's not possible. Should I drop the last asserts?

Gerold103 replied:
Yes. The idea is that you only need those asserts somewhere inside mem_copy_str and mem_copy_bin and other mem methods. You don't need to repeat them before every usage of mem methods.

Gerold103 reviewed (changes requested):
Thanks for the patch!

Gerold103 reviewed (changes requested):
See the previous comment.

Gerold103 reviewed (approved):
Thanks for the fix!

locker commented on src/box/sql/func.c:
```diff
@@ -306,6 +306,11 @@ func_lower_upper(struct sql_context *ctx, int argc, const struct Mem *argv)
+	if (len > SQL_MAX_LENGTH) {
+		diag_set(ClientError, ER_SQL_TOO_BIG_MEM_VALUE);
+		ctx->is_aborted = true;
+		return;
```
This hunk and many others failing on `SQL_MAX_LENGTH` aren't covered by any tests. This is bad because we can't guarantee nothing bad happens if we hit any of those code paths.

Also, I don't like that we check the length in so many places. We can easily forget to add the check to a new places. Actually, I'm not sure you haven't missed a place where the check is required.

I see two ways to improve on this:
1. Check the length only when the mem is set (`mem_copy_bin` and similar functions).
2. Use `size_t` for string length instead of `int` and drop the checks completely or leave them only for user input.

I'd prefer the second way.

Still, we must fix the security issue ASAP so in the scope of this PR I'd fix only those places where we may leak data on invalid input or crash to close the referenced ticket. Refactoring required to make the code neat and clean can be done later.

ImeevMA replied:
Understood. Reworked.

p7nov commented on changelogs/unreleased/ghs-119-too-long-mem-values.md:
```diff
@@ -0,0 +1,3 @@
+## bugfix/sql
+
+* SQL now cannot operate with values larger than SQL_MAX_LENGTH (ghs-119).
```
```suggestion
* SQL now cannot operate with values larger than `SQL_MAX_LENGTH` (ghs-119).
```

ImeevMA replied:
Rewrote the changelog.

ImeevMA commented:
Hi, @Gerold103! On the second review, we decided to go the other way - we plan to abandon the artificial length limit altogether. However, this is quite a big problem and we decided to avoid such big changes in this patch. This patch is only needed to fix the problem, so replacing `int` and `uint32_t` actually did the job thanks to already existing length checks. Could you review this patch once more?

Gerold103 reviewed (approved):
Thanks for the fixes!

locker commented on test/sql-luatest/gh_s119_too_long_mem_values_test.lua:
Nit: We usually use `ghs_XXX` in file names, not `gh_sXXX`:

```
$ find test -name '*ghs*'
test/box-luatest/ghs_16_user_enumeration_test.lua
test/box-luatest/ghs_5_exclude__collation_from_public_access_test.lua
test/box-luatest/ghs_5_fix_auth_check_on_truncate_test.lua
$ find test -name '*gh_s*'
test/sql-luatest/gh_s119_too_long_mem_values_test.lua
```

ImeevMA replied:
Renamed.

ImeevMA commented on src/box/sql/func.c:
```diff
@@ -318,10 +318,9 @@ func_lower_upper(struct sql_context *ctx, int argc, const struct Mem *argv)
-		ucasemap_utf8ToLower(cm, res, len, str, len, &status);
+	size_t size = is_upper ?
+		      ucasemap_utf8ToUpper(cm, res, len, str, len, &status) :
+		      ucasemap_utf8ToLower(cm, res, len, str, len, &status);
```
I reverted this change and added an additional check for the length of the input string. This is a temporary solution, as we need to correctly check the value of the `status` variable.

ImeevMA commented:
@locker fixed. Added more checks and changed types of some variables to `int32_t`.

<details>
<summary>Diff</summary>

```DIff
diff --git a/src/box/sql/func.c b/src/box/sql/func.c
index afed6deb1..af611e9df 100644
--- a/src/box/sql/func.c
+++ b/src/box/sql/func.c
@@ -282,11 +282,17 @@ func_char_length(struct sql_context *ctx, int argc, const struct Mem *argv)
 	if (mem_is_null(arg))
 		return;
 	assert(mem_is_str(arg));
+	if (arg->n > SQL_MAX_LENGTH) {
+		ctx->is_aborted = true;
+		diag_set(ClientError, ER_SQL_EXECUTE, "string or blob too big");
+		return;
+	}
+	int32_t n = arg->n;
 	uint32_t len = 0;
-	size_t offset = 0;
-	while (offset < arg->n) {
+	int32_t offset = 0;
+	while (offset < n) {
 		UChar32 c;
-		U8_NEXT((uint8_t *)arg->z, offset, arg->n, c);
+		U8_NEXT((uint8_t *)arg->z, offset, n, c);
 		++len;
 	}
 	mem_set_uint(ctx->pOut, len);
@@ -427,16 +433,16 @@ func_trim_bin(struct sql_context *ctx, int argc, const struct Mem *argv)
 }
 
 /** Return the position of the last not removed character. */
-static inline size_t
-trim_str_end(const char *str, size_t end, const char *chars, uint8_t *chars_len,
-	     size_t chars_count, int flags)
+static inline int32_t
+trim_str_end(const char *str, int32_t end, const char *chars,
+	     uint8_t *chars_len, size_t chars_count, int flags)
 {
 	if ((flags & TRIM_TRAILING) == 0)
 		return end;
 	while (end > 0) {
 		bool is_trimmed = false;
 		const char *c = chars;
-		size_t len;
+		int32_t len;
 		for (size_t i = 0; i < chars_count && !is_trimmed; ++i) {
 			len = chars_len[i];
 			const char *s = str + end - len;
@@ -452,17 +458,17 @@ trim_str_end(const char *str, size_t end, const char *chars, uint8_t *chars_len,
 }
 
 /** Return the position of the first not removed character. */
-static inline size_t
-trim_str_start(const char *str, size_t end, const char *chars,
+static inline int32_t
+trim_str_start(const char *str, int32_t end, const char *chars,
 	       uint8_t *chars_len, size_t chars_count, int flags)
 {
 	if ((flags & TRIM_LEADING) == 0)
 		return 0;
-	size_t start = 0;
+	int32_t start = 0;
 	while (start < end) {
 		bool is_trimmed = false;
 		const char *c = chars;
-		size_t len;
+		int32_t len;
 		for (size_t i = 0; i < chars_count && !is_trimmed; ++i) {
 			len = chars_len[i];
 			const char *s = str + start;
@@ -487,10 +493,21 @@ func_trim_str(struct sql_context *ctx, int argc, const struct Mem *argv)
 	assert(argc == 2 || (argc == 3 && mem_is_str(&argv[2])));
 	assert(mem_is_str(&argv[0]) && mem_is_uint(&argv[1]));
 	const char *str = argv[0].z;
-	size_t size = argv[0].n;
+	if (argv[0].n > SQL_MAX_LENGTH) {
+		ctx->is_aborted = true;
+		diag_set(ClientError, ER_SQL_EXECUTE, "string or blob too big");
+		return;
+	}
+	int32_t size = argv[0].n;
 	const char *chars;
-	size_t chars_size;
+	int32_t chars_size;
 	if (argc == 3) {
+		if (argv[2].n > SQL_MAX_LENGTH) {
+			ctx->is_aborted = true;
+			diag_set(ClientError, ER_SQL_EXECUTE,
+				 "string or blob too big");
+			return;
+		}
 		chars = argv[2].z;
 		chars_size = argv[2].n;
 	} else {
@@ -508,7 +525,7 @@ func_trim_str(struct sql_context *ctx, int argc, const struct Mem *argv)
 	}
 	size_t chars_count = 0;
 
-	size_t offset = 0;
+	int32_t offset = 0;
 	while (offset < chars_size) {
 		UChar32 c;
 		size_t prev = offset;
@@ -517,10 +534,10 @@ func_trim_str(struct sql_context *ctx, int argc, const struct Mem *argv)
 	}
 
 	uint64_t flags = argv[1].u.u;
-	size_t end = trim_str_end(str, size, chars, chars_len, chars_count,
-				  flags);
-	size_t start = trim_str_start(str, end, chars, chars_len, chars_count,
-				      flags);
+	int32_t end = trim_str_end(str, size, chars, chars_len, chars_count,
+				   flags);
+	int32_t start = trim_str_start(str, end, chars, chars_len, chars_count,
+				       flags);
 	region_truncate(region, svp);
 
 	if (start >= end)
@@ -558,16 +575,21 @@ func_position_characters(struct sql_context *ctx, int argc,
 	if (mem_is_any_null(&argv[0], &argv[1]))
 		return;
 	assert(mem_is_str(&argv[0]) && mem_is_str(&argv[1]));
+	if (argv[0].n > SQL_MAX_LENGTH || argv[1].n > SQL_MAX_LENGTH) {
+		ctx->is_aborted = true;
+		diag_set(ClientError, ER_SQL_EXECUTE, "string or blob too big");
+		return;
+	}
 
 	const char *key = argv[0].z;
 	const char *str = argv[1].z;
-	size_t key_size = argv[0].n;
-	size_t str_size = argv[1].n;
+	int32_t key_size = argv[0].n;
+	int32_t str_size = argv[1].n;
 	if (key_size <= 0)
 		return mem_set_uint(ctx->pOut, 1);
 
-	size_t key_end = 0;
-	size_t str_end = 0;
+	int32_t key_end = 0;
+	int32_t str_end = 0;
 	while (key_end < key_size && str_end < str_size) {
 		UChar32 c;
 		U8_NEXT((uint8_t *)key, key_end, key_size, c);
@@ -580,8 +602,8 @@ func_position_characters(struct sql_context *ctx, int argc,
 	if (coll->cmp(key, key_size, str, str_end, coll) == 0)
 		return mem_set_uint(ctx->pOut, 1);
 
-	size_t i = 2;
-	size_t str_pos = 0;
+	int32_t i = 2;
+	int32_t str_pos = 0;
 	while (str_end < str_size) {
 		UChar32 c;
 		U8_NEXT((uint8_t *)str, str_pos, str_size, c);
@@ -695,10 +717,15 @@ func_substr_characters(struct sql_context *ctx, int argc, const
 	if (mem_is_any_null(&argv[0], &argv[1]))
 		return;
 	assert(mem_is_str(&argv[0]) && mem_is_int(&argv[1]));
+	if (argv[0].n > SQL_MAX_LENGTH) {
+		ctx->is_aborted = true;
+		diag_set(ClientError, ER_SQL_EXECUTE, "string or blob too big");
+		return;
+	}
 
 	const char *str = argv[0].z;
-	size_t pos = 0;
-	size_t end = argv[0].n;
+	int32_t pos = 0;
+	int32_t end = argv[0].n;
 	if (argc == 2) {
 		uint64_t start = mem_is_uint(&argv[1]) && argv[1].u.u > 1 ?
 				 argv[1].u.u - 1 : 0;
@@ -740,7 +767,7 @@ func_substr_characters(struct sql_context *ctx, int argc, const
 	if (pos == end)
 		return mem_set_str_static(ctx->pOut, "", 0);
 
-	size_t cur = pos;
+	int32_t cur = pos;
 	for (uint64_t i = 0; i < length && cur < end; ++i) {
 		UChar32 c;
 		U8_NEXT((uint8_t *)str, cur, end, c);
@@ -932,7 +959,7 @@ func_randomblob(struct sql_context *ctx, int argc, const struct Mem *argv)
 	if (arg->u.u == 0)
 		return mem_set_bin_static(ctx->pOut, "", 0);
 	uint64_t len = arg->u.u;
-	if (len > SIZE_MAX) {
+	if (len > SQL_MAX_LENGTH) {
 		ctx->is_aborted = true;
 		diag_set(ClientError, ER_SQL_EXECUTE, "string or blob too big");
 		return;
@@ -960,7 +987,7 @@ func_zeroblob(struct sql_context *ctx, int argc, const struct Mem *argv)
 	if (arg->u.u == 0)
 		return mem_set_bin_static(ctx->pOut, "", 0);
 	uint64_t len = arg->u.u;
-	if (len > SIZE_MAX) {
+	if (len > SQL_MAX_LENGTH) {
 		ctx->is_aborted = true;
 		diag_set(ClientError, ER_SQL_EXECUTE, "string or blob too big");
 		return;
@@ -1101,9 +1128,16 @@ func_unicode(struct sql_context *ctx, int argc, const struct Mem *argv)
 	assert(mem_is_str(arg));
 	if (arg->n == 0)
 		return mem_set_uint(ctx->pOut, 0);
-	size_t pos = 0;
+	if (arg->n > SQL_MAX_LENGTH) {
+		ctx->is_aborted = true;
+		diag_set(ClientError, ER_SQL_EXECUTE, "string or blob too big");
+		return;
+	}
+
+	int32_t pos = 0;
+	int32_t n = arg->n;
 	UChar32 c;
-	U8_NEXT((uint8_t *)arg->z, pos, arg->n, c);
+	U8_NEXT((uint8_t *)arg->z, pos, n, c);
 	(void)pos;
 	mem_set_uint(ctx->pOut, (uint64_t)c);
 }
@@ -1402,11 +1436,15 @@ static void
 likeFunc(sql_context *context, int argc, const struct Mem *argv)
 {
 	u32 escape = SQL_END_OF_STRING;
-	size_t nPat;
 	assert(argc == 2 || argc == 3);
 	if (mem_is_any_null(&argv[0], &argv[1]))
 		return;
 	assert(mem_is_str(&argv[0]) && mem_is_str(&argv[1]));
+	if (argv[0].n > SQL_MAX_LENGTH || argv[1].n > SQL_MAX_LENGTH) {
+		context->is_aborted = true;
+		diag_set(ClientError, ER_SQL_EXECUTE, "string or blob too big");
+		return;
+	}
 	const char *zB = argv[0].z;
 	const char *zA = argv[1].z;
 	const char *zB_end = zB + argv[0].n;
@@ -1417,8 +1455,8 @@ likeFunc(sql_context *context, int argc, const struct Mem *argv)
 	 * of deep recursion and N*N behavior in
 	 * sql_utf8_pattern_compare().
 	 */
-	nPat = argv[0].n;
-	if (nPat > (size_t)sql_get()->aLimit[SQL_LIMIT_LIKE_PATTERN_LENGTH]) {
+	int32_t nPat = argv[0].n;
+	if (nPat > sql_get()->aLimit[SQL_LIMIT_LIKE_PATTERN_LENGTH]) {
 		diag_set(ClientError, ER_SQL_EXECUTE, "LIKE pattern is too "\
 			 "complex");
 		context->is_aborted = true;
@@ -1428,14 +1466,21 @@ likeFunc(sql_context *context, int argc, const struct Mem *argv)
 	if (argc == 3) {
 		if (mem_is_null(&argv[2]))
 			return;
+		assert(mem_is_str(&argv[2]));
+		if (argv[2].n > SQL_MAX_LENGTH) {
+			context->is_aborted = true;
+			diag_set(ClientError, ER_SQL_EXECUTE,
+				 "string or blob too big");
+			return;
+		}
 		/*
 		 * The escape character string must consist of a
 		 * single UTF-8 character. Otherwise, return an
 		 * error.
 		 */
 		const char *str = argv[2].z;
-		size_t pos = 0;
-		size_t end = argv[2].n;
+		int32_t pos = 0;
+		int32_t end = argv[2].n;
 		U8_NEXT((uint8_t *)str, pos, end, escape);
 		if (pos != end || end == 0) {
 			diag_set(ClientError, ER_SQL_EXECUTE, "ESCAPE "\
diff --git a/test/sql-luatest/suite.ini b/test/sql-luatest/suite.ini
index 6f2975ec4..c7b609bf2 100644
--- a/test/sql-luatest/suite.ini
+++ b/test/sql-luatest/suite.ini
@@ -1,3 +1,4 @@
 [default]
 core = luatest
 description = SQL tests on luatest
+long_run = sql-luatest/ghs_119_too_long_mem_values_test.lua
```

</details>

locker commented:
Cherry-picked to 2.11. Can't cherry-pick to 2.10. @ImeevMA Please prepare a PR that backports this fix to 2.10.
